src/package-linker.js

```diff
@@ -604,49 +670,61 @@ export default class PackageLinker {
       const ref = pkg._reference;
       invariant(ref, 'Package reference is missing');
 
-      // TODO: We are taking the "shortest" ref tree but there may be multiple ref trees with the same length
-      const refTree = ref.requests.map(req => req.parentNames).sort((arr1, arr2) => arr1.length - arr2.length)[0];
-
-      const getLevelDistance = pkgRef => {
-        let minDistance = Infinity;
-        for (const req of pkgRef.requests) {
-          const distance = refTree.length - req.parentNames.length;
-
-          if (distance >= 0 && distance < minDistance && req.parentNames.every((name, idx) => name === refTree[idx])) {
-            minDistance = distance;
-          }
-        }
+      const refTrees = extractRefTrees(ref);
 
-        return minDistance;
-      };
+      if (refTrees.length === 0) {
+        continue;
+      }
 
       for (const peerDepName in peerDeps) {
         const range = peerDeps[peerDepName];
         const meta = peerDepsMeta && peerDepsMeta[peerDepName];
 
         const isOptional = !!(meta && meta.optional);
 
+        const onNotResolved = isOptional
+          ? (refTree, peerError) => {}
+          : (refTree, peerError) => {
+              this.reporter.warn(
+                this.reporter.lang(
+                  peerError,
+                  `${refTree.join(' > ')} > ${pkg.name}@${pkg.version}`,
+                  `${peerDepName}@${range}`,
+                ),
+              );
+            };
+
         const peerPkgs = this.resolver.getAllInfoForPackageName(peerDepName);
 
-        let peerError = 'unmetPeer';
-        let resolvedLevelDistance = Infinity;
-        let resolvedPeerPkg;
-        for (const peerPkg of peerPkgs) {
-          const peerPkgRef = peerPkg._reference;
-          if (!(peerPkgRef && peerPkgRef.patterns)) {
-            continue;
-          }
-          const levelDistance = getLevelDistance(peerPkgRef);
-          if (isFinite(levelDistance) && levelDistance < resolvedLevelDistance) {
-            if (this._satisfiesPeerDependency(range, peerPkgRef.version)) {
-              resolvedLevelDistance = levelDistance;
-              resolvedPeerPkg = peerPkgRef;
-            } else {
-              peerError = 'incorrectPeer';
+        const refTreesStatus = [];
```

Comment:
We start by building an array containing the results on a per reference tree basis.
The first one would be the first reference tree found which (by construction) would be the first defined of the following:

* dependency defined at the root level
* first package (according to the platform's sorting conventions) which defines a dependency on that package